File: Algorithm.Python/BrokerageModelAlgorithm.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

### <summary>
### Demonstrate the usage of the BrokerageModel property to help improve backtesting
### accuracy through simulation of a specific brokerage's rules around restrictions
### on submitting orders as well as fee structure.
### </summary>
### <meta name="tag" content="trading and orders" />
### <meta name="tag" content="brokerage models" />
class BrokerageModelAlgorithm(QCAlgorithm):
    def initialize(self):

        self.set_cash(100000)            # Set Strategy Cash
        self.set_start_date(2013,10,7)    # Set Start Date
        self.set_end_date(2013,10,11)     # Set End Date
        self.add_equity("SPY", Resolution.SECOND)
        
        # there's two ways to set your brokerage model. The easiest would be to call
        # self.set_brokerage_model( BrokerageName ) # BrokerageName is an enum
        # self.set_brokerage_model(BrokerageName.INTERACTIVE_BROKERS_BROKERAGE)
        # self.set_brokerage_model(BrokerageName.DEFAULT)

        # the other way is to call SetBrokerageModel( IBrokerageModel ) with your
        # own custom model. I've defined a simple extension to the default brokerage
        # model to take into account a requirement to maintain 500 cash in the account at all times
        self.set_brokerage_model(MinimumAccountBalanceBrokerageModel(self,500.00))
        self.last = 1   
        
    def on_data(self, slice):
        # Simple buy and hold template
        if not self.portfolio.invested:
            self.set_holdings("SPY", self.last)
            if self.portfolio["SPY"].quantity == 0:
                # each time we fail to purchase we'll decrease our set holdings percentage
                self.debug(str(self.time) + " - Failed to purchase stock")
                self.last *= 0.95
            else:
                self.debug("{} - Purchased Stock @ SetHoldings( {} )".format(self.time, self.last))


class MinimumAccountBalanceBrokerageModel(DefaultBrokerageModel):
    '''Custom brokerage model that requires clients to maintain a minimum cash balance'''
    def __init__(self, algorithm, minimum_account_balance):
        self.algorithm = algorithm
        self.minimum_account_balance = minimum_account_balance
    
    def can_submit_order(self,security, order, message):
        '''Prevent orders which would bring the account below a minimum cash balance'''
        message = None
        # we want to model brokerage requirement of minimum_account_balance cash value in account
        order_cost = order.get_value(security)
        cash = self.algorithm.portfolio.cash
        cash_after_order = cash - order_cost
        if cash_after_order < self.minimum_account_balance:
            # return a message describing why we're not allowing this order
            message = BrokerageMessageEvent(BrokerageMessageType.WARNING, "InsufficientRemainingCapital", "Account must maintain a minimum of ${0} USD at all times. Order ID: {1}".format(self.minimum_account_balance, order.id))
            self.algorithm.error(str(message))
            return False
        return True

File: Algorithm.CSharp/BrokerageModelAlgorithm.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using QuantConnect.Brokerages;
using QuantConnect.Data.Market;
using QuantConnect.Orders;
using QuantConnect.Securities;

namespace QuantConnect.Algorithm.CSharp
{
    /// <summary>
    /// Demonstrate the usage of the BrokerageModel property to help improve backtesting
    /// accuracy through simulation of a specific brokerage's rules around restrictions
    /// on submitting orders as well as fee structure.
    /// </summary>
    /// <meta name="tag" content="trading and orders" />
    /// <meta name="tag" content="brokerage models" />
    public class BrokerageModelAlgorithm : QCAlgorithm
    {
        /// <summary>
        /// Initialize the data and resolution required, as well as the cash and start-end dates for your algorithm. All algorithms must be initialized.
        /// </summary>
        public override void Initialize()
        {
            SetStartDate(2013, 10, 07);  //Set Start Date
            SetEndDate(2013, 10, 11);    //Set End Date
            SetCash(100000);             //Set Strategy Cash
            // Find more symbols here: http://quantconnect.com/data
            AddSecurity(SecurityType.Equity, "SPY", Resolution.Second);

            // there's two ways to set your brokerage model. The easiest would be to call
            // SetBrokerageModel( BrokerageName ); // BrokerageName is an enum
            //SetBrokerageModel(BrokerageName.InteractiveBrokersBrokerage);
            //SetBrokerageModel(BrokerageName.Default);

            // the other way is to call SetBrokerageModel( IBrokerageModel ) with your
            // own custom model. I've defined a simple extension to the default brokerage
            // model to take into account a requirement to maintain 500 cash in the account
            // at all times

            SetBrokerageModel(new MinimumAccountBalanceBrokerageModel(this, 500.00m));
        }

        private decimal _last = 1.0m;

        /// <summary>
        /// OnData event is the primary entry point for your algorithm. Each new data point will be pumped in here.
        /// </summary>
        /// <param name="data">TradeBars IDictionary object with your stock data</param>
        public void OnData(TradeBars data)
        {
            if (!Portfolio.Invested)
            {
                //fails first several times, we'll keep decrementing until it succeeds
                SetHoldings("SPY", _last);
                if (Portfolio["SPY"].Quantity == 0)
                {
                    // each time we fail to purchase we'll decrease our set holdings percentage
                    Debug(Time + " - Failed to purchase stock");
                    _last *= 0.95m;
                }
                else
                {
                    Debug(Time + " - Purchased Stock @ SetHoldings( " + _last + " )");
                }
            }
        }

        /// <summary>
        /// Custom brokerage model that requires clients to maintain a minimum cash balance
        /// </summary>
        class MinimumAccountBalanceBrokerageModel : DefaultBrokerageModel
        {
            private readonly QCAlgorithm _algorithm;
            private readonly decimal _minimumAccountBalance;

            public MinimumAccountBalanceBrokerageModel(QCAlgorithm algorithm, decimal minimumAccountBalance)
            {
                _algorithm = algorithm;
                _minimumAccountBalance = minimumAccountBalance;
            }

            /// <summary>
            /// Prevent orders which would bring the account below a minimum cash balance
            /// </summary>
            public override bool CanSubmitOrder(Security security, Order order, out BrokerageMessageEvent message)
            {
                message = null;

                // we want to model brokerage requirement of _minimumAccountBalance cash value in account

                var orderCost = order.GetValue(security);
                var cash = _algorithm.Portfolio.Cash;
                var cashAfterOrder = cash - orderCost;
                if (cashAfterOrder < _minimumAccountBalance)
                {
                    // return a message describing why we're not allowing this order
                    message = new BrokerageMessageEvent(BrokerageMessageType.Warning, "InsufficientRemainingCapital",
                        $"Account must maintain a minimum of ${_minimumAccountBalance.ToStringInvariant()} USD at all times. " +
                        $"Order ID: {order.Id.ToStringInvariant()}"
                    );
                    return false;
                }
                return true;
            }
        }
    }
}